### ISSUE TITLE
Silent bad codegen when passing and comparing ushort args

### ISSUE THREAD

**Issue by erozenfeld** [dotnet/coreclr#11508]
Repro:

Compile Program.cs below:

.\Tools\dotnetcli\dotnet.exe Tools\csc.exe /o+ /out:Program.exe Program.cs /r:.\bin\tests\Windows_NT.x64.Checked\Tests\Core_Root\System.Private.CoreLib.dll /r:.\bin\tests\Windows_NT.x64.Checked\Tests\Core_Root\System.Console.dll /r:.\bin\tests\Windows_NT.x64.Checked\Tests\Core_Root\System.Runtime.dll

Program.cs:
```
using System;
using System.Collections.Generic;
using System.Runtime.CompilerServices;

namespace TestApp
{
    public struct StructWithValue : IEquatable<StructWithValue>
    {
        private ushort value;

        public StructWithValue(ushort v)
        {
            value = v;
        }


        public bool Equals(StructWithValue other)
        {
            if (value.Equals (other.value))
                return true;
            else
                return false;
        }
    }
    class Program
    {
        [MethodImpl(MethodImplOptions.NoOptimization)]
        static void Main(string[] args)
        {
            var comparer = EqualityComparer<StructWithValue>.Default;
            
            for (ushort i = short.MaxValue - 1; i <= 32770; i++)
            {
                Console.WriteLine ("i = {0}", i);
                var a = new StructWithValue(i);
                var b = new StructWithValue(i);

                if (comparer.Equals(a, b))
                    continue;

                // The value of 32768 will show this message and end the loop
                // Every value > 32767 is affected...

                Console.WriteLine("Problem starts at {0}", i);

                break;

            }

        }

    }

}

```
.\bin\tests\Windows_NT.x64.Checked\Tests\Core_Root\CoreRun.exe Program.exe
outputs
i = 32766
i = 32767
i = 32768
Problem starts at 32768

**Comment by erozenfeld** [comment]
Codegen for passing the args to `comparer.Equals(a, b)`:
```
       488B4DE8             mov      rcx, gword ptr [rbp-18H]
       480FBF55D8           movsx    rdx, word  ptr [rbp-28H]
       4C0FBF45D0           movsx    r8, word  ptr [rbp-30H]
       488B45E8             mov      rax, gword ptr [rbp-18H]
       488B00               mov      rax, qword ptr [rax]
       488B4048             mov      rax, qword ptr [rax+72]
       FF5020               call     qword ptr [rax+32]System.Collections.Generic.EqualityComparer`1[StructWithValue][TestApp.StructWithValue]:Equals(struct,struct):bool:this
```
Note the sign extension for both arguments.

Codegen for System.Collections.Generic.GenericEqualityComparer`1[StructWithValue][TestApp.StructWithValue]:Equals(struct,struct):bool:this

```
G_M34379_IG01:

G_M34379_IG02:
       410FB7C0             movzx    rax, r8w
       0FB7C0               movzx    rax, ax
       3BD0                 cmp      edx, eax
       7507                 jne      SHORT G_M34379_IG03
       B801000000           mov      eax, 1
       EB02                 jmp      SHORT G_M34379_IG04

G_M34379_IG03:
       33C0                 xor      eax, eax

G_M34379_IG04:
       C3                   ret      
```
Zero extension is done only on one of the parameters.

**Comment by RussKeldorph** [comment]
@dotnet/jit-contrib

**Comment by karelz** [comment]
Post-ZBB as it was created very late (on ZBB date). Keeping it in 2.0.

**Comment by mikedn** [comment]
@pgavlin In case you haven't looked at this yet the problem appears to be in morph. At some point we have
```
[000057] ------------                   |  /--*  FIELD     char   value
[000056] L-----------                   |  |  \--*  ADDR      byref 
[000055] ------------                   |  |     \--*  LCL_VAR   struct V05 tmp2         
[000092] ---XG-------                   \--*  EQ        int   
[000090] *--XG-------                      \--*  IND       char  
[000086] ----G-------                         \--*  ADDR      byref 
[000087] ----G-------                            \--*  FIELD     char   value
[000088] L-----------                               \--*  ADDR      byref 
[000089] ------------                                  \--*  LCL_VAR   struct V01 arg1         
```
The top field gets transformed into a lclvar and then morph inserts a cast. The bottom indir gets transformed first into ind/addr/lclvar and then into lclvar. Now it's too late for the cast to be inserted.

**Comment by russellhadley** [comment]
@erozenfeld, just to keep context in morph.

**Comment by RussKeldorph** [comment]
@pgavlin Should this be closed by #11531?

**Comment by pgavlin** [comment]
Yes.
